pydropsonde/circles.py

```diff
@@ -15,6 +23,172 @@ class Circle:
+        circle_diameter = np.nanmean(c_r) * 2
+
+        xc = [None] * len(x_coor.T)
+        yc = [None] * len(y_coor.T)
+
+        xc = np.mean(x_coor, axis=0)
+        yc = np.mean(y_coor, axis=0)
+
+        delta_x = x_coor - xc  # *111*1000 # difference of sonde long from mean long
+        delta_y = y_coor - yc  # *111*1000 # difference of sonde lat from mean lat
+
+        new_vars = dict(
+            circle_flight_altitude=self.circle_ds["aircraft_geopotential_altitude"]
+            .mean()
+            .values,
+            circle_time=self.circle_ds["launch_time"].mean().values,
+            circle_lon=circle_x,
+            circle_lat=circle_y,
+            circle_diameter=circle_diameter,
+            dx=(["sonde_id", "alt"], delta_x.values),
+            dy=(["sonde_id", "alt"], delta_y.values),
+        )
+        self.circle_ds = self.circle_ds.assign(new_vars)
+
+        return self
+
+    @staticmethod
+    def fit2d(x, y, u, var=None):
+        # Function logic goes here
+        # No need for self here since it's now a static method
+        a = np.stack([np.ones_like(x), x, y], axis=-1)
+
+        invalid = np.isnan(u) | np.isnan(x) | np.isnan(y)
+        u_cal = np.where(invalid, 0, u)
+        a[invalid] = 0
+
+        a_inv = np.linalg.pinv(a)
+        intercept, dudx, dudy = np.einsum("...rm,...m->r...", a_inv, u_cal)
+
+        return intercept, dudx, dudy
+
+    def fit2d_xr(self, x, y, u, sonde_dim="sonde_id"):
+        # Apply the static method fit2d without passing self
+        return xr.apply_ufunc(
+            self.__class__.fit2d,  # Call the static method without passing `self`
+            x,
+            y,
+            u,
+            input_core_dims=[
+                [sonde_dim],
+                [sonde_dim],
+                [sonde_dim],
+            ],  # Specify input dims
+            output_core_dims=[(), (), ()],  # Output dimensions as scalars
+        )
+
+    def apply_fit2d(self):
+        # Loop over the parameters to apply fit2d_xr to each of them
+        for par in tqdm.tqdm(["u", "v", "q", "ta", "p"]):
+            varnames = [par + "0", "d" + par + "dx", "d" + par + "dy"]
+
+            # Apply fit2d_xr to each variable, and assign the result to circle_ds
+            results = self.fit2d_xr(
+                x=self.circle_ds.dx,
+                y=self.circle_ds.dy,
+                u=self.circle_ds[par],
+                sonde_dim="sonde_id",
+            )
+
+            # Assign the results using the varnames list for the output variables
+            self.circle_ds = self.circle_ds.assign(
+                {varname: result for varname, result in zip(varnames, results)}
+            )
+
+        return self
+
+    def get_div_and_vor(self):
+        D = self.circle_ds.dudx + self.circle_ds.dvdy
+        vor = self.circle_ds.dvdx - self.circle_ds.dudy
+
+        self.circle_ds = self.circle_ds.assign(
+            dict(div=(["alt"], D.values), vor=(["alt"], vor.values))
```

Comment:
maybe it would be good to add attributes here as well, explaining what the variables are and what unit they have